packages/ra-core/src/dataProvider/useDataProviderWithDeclarativeSideEffects.ts

```diff
@@ -0,0 +1,73 @@
+import {
+    useNotify,
+    useRedirect,
+    useRefresh,
+    useUnselectAll,
+} from '../sideEffect';
+import useDataProvider, { DataProviderHookFunction } from './useDataProvider';
+import { useCallback } from 'react';
+
+const useDataProviderWithDeclarativeSideEffects = (): DataProviderHookFunction => {
```

Comment:
I'd add a jsDoc block explaining that this is for backwards compatibility (to be removed in a future major version)